File: ios/PhoneCheck/UpdateChecker.swift
import Foundation

struct GitHubRelease: Decodable {
    let tagName: String
    let htmlURL: String

    enum CodingKeys: String, CodingKey {
        case tagName = "tag_name"
        case htmlURL = "html_url"
    }
}

struct UpdateCheckResult {
    let message: String
    let releaseURL: URL?
}

enum UpdateChecker {
    private static let releaseURL = URL(string: "https://api.github.com/repos/catshub/phone-check/releases/latest")!

    static func check() async -> UpdateCheckResult {
        var request = URLRequest(url: releaseURL)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("phone-check-ios", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return UpdateCheckResult(message: "GitHub 没有发布信息", releaseURL: nil)
            }

            let release = try JSONDecoder().decode(GitHubRelease.self, from: data)
            let latest = release.tagName.replacingOccurrences(of: "v", with: "")
            let current = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
            let message = isNewer(latest, current)
                ? "发现新版本 \(latest)"
                : "已是最新版本"
            return UpdateCheckResult(message: message, releaseURL: URL(string: release.htmlURL))
        } catch {
            return UpdateCheckResult(message: "检查更新失败", releaseURL: nil)
        }
    }

    private static func isNewer(_ latest: String, _ current: String) -> Bool {
        let latestParts = latest.split(separator: ".").map { Int($0) ?? 0 }
        let currentParts = current.split(separator: ".").map { Int($0) ?? 0 }
        let count = max(latestParts.count, currentParts.count)

        for index in 0..<count {
            let left = index < latestParts.count ? latestParts[index] : 0
            let right = index < currentParts.count ? currentParts[index] : 0
            if left != right { return left > right }
        }
        return false
    }
}

File: app/src/main/java/com/xq/phonecheck/update/UpdateChecker.kt
package com.xq.phonecheck.update

import android.content.Context
import android.os.Handler
import android.os.Looper
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

sealed class UpdateResult {
    object UpToDate : UpdateResult()
    data class UpdateAvailable(
        val version: String,
        val url: String,
        val apkUrl: String? = null
    ) : UpdateResult()
    data class Failure(val message: String) : UpdateResult()
}

object UpdateChecker {
    private const val RELEASE_URL = "https://api.github.com/repos/catshub/phone-check/releases/latest"
    private const val USER_AGENT = "phone-check-android"

    fun check(context: Context, callback: (UpdateResult) -> Unit) {
        val appContext = context.applicationContext
        val mainHandler = Handler(Looper.getMainLooper())

        Thread {
            val result = runCatching {
                val connection = URL(RELEASE_URL).openConnection() as HttpURLConnection
                connection.connectTimeout = 8_000
                connection.readTimeout = 8_000
                connection.setRequestProperty("Accept", "application/vnd.github+json")
                connection.setRequestProperty("User-Agent", USER_AGENT)

                val responseCode = connection.responseCode
                if (responseCode != HttpURLConnection.HTTP_OK) {
                    UpdateResult.Failure("GitHub 返回 $responseCode")
                } else {
                    val body = connection.inputStream.bufferedReader().use { it.readText() }
                    val json = JSONObject(body)
                    val latestTag = json.optString("tag_name")
                    val latestVersion = latestTag.removePrefix("v")
                    val releaseUrl = json.optString("html_url")
                    val apkUrl = findApkAsset(json, latestTag)
                    val currentVersion = appContext.packageManager
                        .getPackageInfo(appContext.packageName, 0)
                        .versionName.orEmpty()

                    if (latestVersion.isBlank() || releaseUrl.isBlank()) {
                        UpdateResult.Failure("GitHub 没有发布信息")
                    } else if (isNewer(latestVersion, currentVersion)) {
                        UpdateResult.UpdateAvailable(latestVersion, releaseUrl, apkUrl)
                    } else {
                        UpdateResult.UpToDate
                    }
                }
            }.getOrElse {
                UpdateResult.Failure(it.message ?: "检查更新失败")
            }

            mainHandler.post { callback(result) }
        }.start()
    }

    internal fun isNewer(latest: String, current: String): Boolean {
        val latestParts = latest.removePrefix("v").split(".").map { it.toIntOrNull() ?: 0 }
        val currentParts = current.removePrefix("v").split(".").map { it.toIntOrNull() ?: 0 }
        val count = maxOf(latestParts.size, currentParts.size)

        for (index in 0 until count) {
            val latestValue = latestParts.getOrElse(index) { 0 }
            val currentValue = currentParts.getOrElse(index) { 0 }
            if (latestValue != currentValue) return latestValue > currentValue
        }
        return false
    }

    private fun findApkAsset(release: JSONObject, latestTag: String): String? {
        val assets = release.optJSONArray("assets") ?: return null
        var firstApk: String? = null
        var preferredApk: String? = null

        for (index in 0 until assets.length()) {
            val asset = assets.optJSONObject(index) ?: continue
            val name = asset.optString("name")
            val url = asset.optString("browser_download_url")
            if (!name.endsWith(".apk") || url.isBlank()) continue

            if (firstApk == null) firstApk = url
            if (name == "phone-check-${latestTag.removePrefix("v")}-debug.apk") {
                preferredApk = url
                break
            }
        }

        return preferredApk ?: firstApk
    }
}
